### PR TITLE
Check for block parent before verification.

### PR DESCRIPTION
**Pull request by arkpar**
* Before verifying block we want to make sure we have the parent. The sync module needs to detect these kind of errors and handle them in a special way.
* Sync now tracks request all canceled requests. This should prevent "Unexpected response" errors and peers getting banned.

### CODE REVIEW

File: core/client/src/client.rs [hunk 1]
```diff
@@ -1302,6 +1302,29 @@ impl<B, E, Block, RA> consensus::BlockImport<Block> for Client<B, E, Block, RA>
...
 	}
+
+	/// Check block preconditions.
+	fn check_block(
+		&self,
+		hash: Block::Hash,
+		parent_hash: Block::Hash,
+	) -> Result<ImportResult, Self::Error> {
+		match self.backend.blockchain().status(BlockId::Hash(parent_hash))
+			.map_err(|e| ConsensusErrorKind::ClientImport(e.to_string()).into())
```
Comment (bkchr) [inline core/client/src/client.rs]: Can we not just use `?` here at the end?

File: core/client/src/client.rs [hunk 2]
```diff
@@ -1302,6 +1302,29 @@ impl<B, E, Block, RA> consensus::BlockImport<Block> for Client<B, E, Block, RA>
...
+		{
+			Ok(blockchain::BlockStatus::InChain) => {},
+			Ok(blockchain::BlockStatus::Unknown) => return Ok(ImportResult::UnknownParent),
+			Err(e) => return Err(e),
+		}
+		match self.backend.blockchain().status(BlockId::Hash(hash))
```
Comment (bkchr) [inline core/client/src/client.rs]: Same.

File: core/network/src/protocol.rs [hunk 3]
```diff
@@ -433,72 +417,24 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
...
 				.network_chan
-				.send(NetworkMsg::ReportPeer(who, Severity::Bad("Unexpected packet received from peer".to_string())));
-			return None;
-		};
-
-		Some(request)
+				.send(NetworkMsg::ReportPeer(who, Severity::Bad("Unexpected response packet received from peer".to_string())));
+			None
+		} else {
+			None
```
Comment (bkchr) [inline core/network/src/protocol.rs]: We can remove the `else` branch and just put `None` at the end of the function

File: core/client/src/client.rs (hunk 1 above)
Comment (arkpar) [reply core/client/src/client.rs]: Not without specifying the type for `into`, which I did not bother to.

**Pull request merged**